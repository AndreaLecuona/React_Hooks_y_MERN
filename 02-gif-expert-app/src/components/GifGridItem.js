import React from 'react';
import Proptypes from "prop-types";

export const GifGridItem = ( { id, title, url } ) => {

    // console.log(id, title, url);

    return (
        <div className="card animate__animated animate__fadeIn">
            <img src={ url } alt={ title } />
            <p> { title } </p>
        </div>
    )
};

GifGridItem.proptypes = {
    title: Proptypes.string.isRequired,
    url: Proptypes.string.isRequired
};
