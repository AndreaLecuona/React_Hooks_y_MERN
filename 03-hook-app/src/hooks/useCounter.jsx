import { useState } from "react"


//centraliza la lógica de "contador" para exportarla y reutilizarla en todos los componentes que la necesiten (y asi evitar repetir codigo)
export const useCounter = ( initialState = 1 ) => {

    const [counter, setCounter] = useState(initialState);

    const increment = () => {
        setCounter( counter + 1 );
    }

    const decrement = () => {
        setCounter( counter - 1 ); 
    }

    const reset = () => {
        setCounter( initialState );
    };

    //flexibilizo aun más, dandole la posibilidad de modificar en cuánto debe crecer o decrecer
    // const increment = ( factor = 1) => {
    //     setCounter( counter + factor );
    // }

    // const decrement = ( factor = 1) => {
    //     setCounter( counter - factor ); 
    // }

    // const reset = () => {
    //     setCounter( initialState );
    // };

    return{
        counter,
        increment,
        decrement,
        reset
    };
}
