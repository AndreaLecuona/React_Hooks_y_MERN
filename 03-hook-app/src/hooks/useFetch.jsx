import { useEffect, useRef, useState } from "react";

//custom hook que hace fetch de los datos de un endpoint que se le pasa como argumento
export const useFetch = (url) => {


    //USEREF para que este custom hook sea más seguro y dé menos lugar a errores
    //hacemos que isMounted (el useRef) mantenga la referencia siempre que este hook esté vivo o cuando el componente que lo usa siga montado.
    //cuando cambiemos los valores del isMounted, no queremos que vuelva a renderizarse el componente
    const isMounted = useRef(true);

    useEffect(() => {
        //no tiene efecto, solo aprovechamos la limpieza

        return () => {
            isMounted.current = false;
        }
    }, []); //se ejecuta una sola vez onpage load



    const [state, setState] = useState({ data: null, loading: true, error: null });

    useEffect(() => {

        setState({ data: null, loading: true, error: null }); //para que vuelva a aparecer el loading entre fetchs de frases

        fetch(url)
            .then(resp => resp.json())
            .then(data => {

                //si el componente esta montado DE MANERA SEGURA, entonces continuamos la ejecución que devuelve la data. Si no, no hacemos nada y evitamos un error que crashea el programa
                if( isMounted.current ){

                    setState({
                        loading: false,
                        error: null,
                        data
                    })

                } else {
                    console.log('setState no se ejecutó');
                }


            })
        .catch( () => {
            setState({
                data: null,
                loading: false,
                error: 'No pudo obtenerse la data'
            })
        })

    }, [url]);

    return state;

};
