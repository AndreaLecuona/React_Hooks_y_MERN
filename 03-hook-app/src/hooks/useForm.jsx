import { useState } from "react";

//extraemos la logica de las forms, que de otra forma se repetiría a lo largo de las secciones de una página
export const useForm = ( initialState = {} ) => {

    const [values, setValues] = useState(initialState);

    const reset = () => {
        setValues( initialState );
    }

    const handleInputChange = ( { target } ) => {

        setValues( {
            ...values,
            [ target.name ]: target.value
        });
    };

    return [ values, handleInputChange, reset ];

};
