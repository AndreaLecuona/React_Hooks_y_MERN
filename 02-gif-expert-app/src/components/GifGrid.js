import React from 'react';
import Proptypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs';
// import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ( { categoria } ) => {

    // const [images, setImages] = useState([]);
    
    const { data: images, loading } = useFetchGifs( categoria );

    // useEffect( () => {
    //     getGifs( categoria ).then( imgs => setImages( imgs )); //al modularizar getGifs, hubo que convertirlo en async y por ende en promesa
    // }, [ categoria ]); //dependencia a la categoria: si cambia la categoria vuelve a hacer la peticion a la api

    return (
        <>
            <h3 className="animate__animated animate__fadeIn">"{ categoria }"</h3>

            { loading && <p className="animate__animated animate__flash">Cargando</p> }
            
            <div className="card-grid">
                {
                    images.map( img =>
                        <GifGridItem 
                            key={img.id}

                            // img={ img }
                            { ...img }
                        />
                    )
                }
            </div>
        </>
    )
};

GifGrid.proptypes = {
    categoria: Proptypes.string.isRequired
};

// SIN DESTRUCTURING

// <div>
// <h3>{ category }</h3>
// <ol>
//     {
//         images.map( image =>
//             <li key={image.id}>{image.title}</li>)
//     }
// </ol>
// </div
