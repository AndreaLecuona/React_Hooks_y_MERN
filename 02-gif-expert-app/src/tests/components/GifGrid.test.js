import React from 'react';
import '@testing-library/jest-dom';
import { GifGrid } from "../../components/GifGrid";
import {shallow} from 'enzyme';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Pruebas en el componente GifGrid', () => {
    
    const categoria = '101 Dalmatas';

    test('Debe renderizarse correctamente', () => {

        //simulamos cuál sería el resultado de llamar al custom hook en un estado inicial
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });
       
        const wrapper = shallow( <GifGrid categoria={ categoria }/> );

        expect( wrapper ).toMatchSnapshot();

    });

    test('Debe mostrar items cuando recibe las imágenes del useFetchGif', () => {
       
        //debemos simular el llamado al custom hook. 
        //Para eso primero se importa el hook, luego se setea jest.mock('path del hook')

        //inventamos una data en el formato que podría recibirla el componente
        const gifs= [{
            id: '123',
            url: 'https://images.app.goo.gl/R4F9rcT6Nhi1ELgo6',
            title: 'Ejemplo'
        }];

        //simulamos el llamado al hook y le ponemos la data inventada como supuesta respuesta
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        //capturo el componente luego del llamado al hook
        const wrapper = shallow( <GifGrid categoria={ categoria }/> );

        // expect( wrapper ).toMatchSnapshot(); //sabemos que funciona por test anterior

        //testeamos que el parrafo 'cargando' no exista
        expect( wrapper.find('p').exists() ).toBe( false );

        //testeamos que se genere un componente GifGridItem
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length );
    });

});