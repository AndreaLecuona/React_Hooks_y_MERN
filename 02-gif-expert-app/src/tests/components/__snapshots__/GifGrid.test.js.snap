// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente GifGrid Debe renderizarse correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    "
    101 Dalmatas
    "
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Cargando
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
